### PR TITLE
support collection level rate limit

### PR DESCRIPTION
**Pull request by weiliu1031**
Signed-off-by: Wei Liu <wei.liu@zilliz.com>
issue #22385

### CODE REVIEW

File: internal/proxy/impl.go [hunk 1]
```diff
@@ -4327,25 +4325,21 @@ func (node *Proxy) SetRates(ctx context.Context, request *proxypb.SetRatesReques
...
 	}
 
-	err := node.multiRateLimiter.globalRateLimiter.setRates(request.GetRates())
+	collectionRates := make(map[string]*proxypb.CollectionRate)
+	for _, rate := range request.GetRates() {
+		collectionName, err := globalMetaCache.GetCollectionName(ctx, rate.GetCollection())
+		if err != nil {
+			log.Warn("found rates with collection name not found",
+				zap.Int64("collectionID", rate.GetCollection()))
+		}
```
Comment (jiaoew1991) [inline internal/proxy/impl.go:4334]: continue when error?

File: internal/proxy/multi_rate_limiter.go [hunk 2]
```diff
@@ -49,80 +51,83 @@ func GetQuotaErrorString(errCode commonpb.ErrorCode) string {
 // MultiRateLimiter includes multilevel rate limiters, such as global rateLimiter,
 // collection level rateLimiter and so on. It also implements Limiter interface.
 type MultiRateLimiter struct {
-	globalRateLimiter *rateLimiter
 	// TODO: add collection level rateLimiter
-	quotaStatesMu sync.RWMutex
-	quotaStates   map[milvuspb.QuotaState]commonpb.ErrorCode
+	quotaStatesMu      sync.RWMutex
+	collectionLimiters map[string]*rateLimiter
```
Comment (jiaoew1991) [inline internal/proxy/multi_rate_limiter.go:56]: why not use collectionID as the key?

File: internal/proxy/multi_rate_limiter.go [hunk 3]
```diff
@@ -49,80 +51,83 @@ func GetQuotaErrorString(errCode commonpb.ErrorCode) string {
...
-		states = append(states, k)
-		reasons = append(reasons, GetQuotaErrorString(v))
+	serviceStates := make(map[milvuspb.QuotaState]string)
+
+	for _, limiter := range m.collectionLimiters {
+		limiter.quotaStates.Range(func(key, value interface{}) bool {
+			serviceStates[key.(milvuspb.QuotaState)] = GetQuotaErrorString(value.(commonpb.ErrorCode))
+			return true
+		})
 	}
```
Comment (jiaoew1991) [inline internal/proxy/multi_rate_limiter.go:99]: Will the values in `serviceStates` be repeatedly overwritten?

File: internal/datanode/metrics_info.go [hunk 4]
```diff
@@ -48,6 +48,17 @@ func (node *DataNode) getQuotaMetrics() (*metricsinfo.DataNodeQuotaMetrics, erro
 	if err != nil {
 		return nil, err
 	}
+
+	getAllCollections := func() []int64 {
+		collectionSet := typeutil.UniqueSet{}
+		node.flowgraphManager.flowgraphs.Range(func(key, value interface{}) bool {
```
Comment (jiaoew1991) [inline internal/datanode/metrics_info.go:54]: use `any` other than `interface{}`

File: internal/rootcoord/quota_center.go [hunk 5]
```diff
@@ -83,13 +84,15 @@ type QuotaCenter struct {
...
-	currentRates map[internalpb.RateType]Limit
-	quotaStates  map[milvuspb.QuotaState]commonpb.ErrorCode
+	queryNodeMetrics    map[UniqueID]*metricsinfo.QueryNodeQuotaMetrics
+	dataNodeMetrics     map[UniqueID]*metricsinfo.DataNodeQuotaMetrics
+	proxyMetrics        map[UniqueID]*metricsinfo.ProxyQuotaMetrics
+	dataCoordMetrics    *metricsinfo.DataCoordQuotaMetrics
+	readableCollections []int64
+	writableCollections []int64
+
+	currentRates map[int64]map[internalpb.RateType]Limit
```
Comment (jiaoew1991) [inline internal/rootcoord/quota_center.go:94]: The two-layer map structure is very complex, define a new type to represent `map[internalpb.RateType]Limit`.

File: internal/proxy/multi_rate_limiter.go (hunk 2 above)
Comment (weiliu1031) [reply internal/proxy/multi_rate_limiter.go:56]: > why not use collectionID as the key?

cause request  in proxy only contains collection name, check with collection_name is a lower cost choice

File: internal/proxy/multi_rate_limiter.go [hunk 6]
```diff
@@ -49,80 +50,98 @@ func GetQuotaErrorString(errCode commonpb.ErrorCode) string {
...
-	m.globalRateLimiter = newRateLimiter()
+	m := &MultiRateLimiter{
+		collectionLimiters: make(map[int64]*rateLimiter, 0),
+	}
 	return m
 }
 
 // Check checks if request would be limited or denied.
-func (m *MultiRateLimiter) Check(rt internalpb.RateType, n int) commonpb.ErrorCode {
+func (m *MultiRateLimiter) Check(collectionID int64, rt internalpb.RateType, n int) commonpb.ErrorCode {
```
Comment (jiaoew1991) [inline internal/proxy/multi_rate_limiter.go:67]: Returning the error redefined by merr is better.

File: internal/proxy/rate_limit_interceptor.go [hunk 7]
```diff
@@ -22,65 +22,116 @@ import (
...
-			if rsp != nil {
-				return rsp, nil
+
+		switch request := req.(type) {
+		case *milvuspb.FlushRequest:
+			collectionNames := request.GetCollectionNames()
+			for _, collectionName := range collectionNames {
+				collectionID, err = globalMetaCache.GetCollectionID(context.TODO(), collectionName)
+				if err != nil {
+					log.Debug("skip limit check",
```
Comment (jiaoew1991) [inline internal/proxy/rate_limit_interceptor.go:49]: Is the log level set to debug correct?

File: internal/proxy/multi_rate_limiter.go [hunk 8]
```diff
@@ -138,20 +157,53 @@ func (rl *rateLimiter) limit(rt internalpb.RateType, n int) (bool, float64) {
...
 		}
 	}
-	// rl.printRates(rates)
+
+	for i := 0; i < len(collectionRate.States); i++ {
+		rl.quotaStates.Insert(collectionRate.States[i], collectionRate.Codes[i])
+	}
+	log.RatedInfo(30, "current collection rates in proxy",
+		zap.Int64("collectionID", collectionRate.GetCollection()),
+		zap.Any("rates", collectionRate.GetRates()),
```
Comment (bigsheeper) [inline internal/proxy/multi_rate_limiter.go:178]: Directly printing rates here can be very annoying, like this:
```
[rates="[{\"rt\":9,\"r\":1.7976931348623157e+308},{\"rt\":5,\"r\":6286213.12},{\"rt\":6,\"r\":523851.0933333333},{\"rt\":7,\"r\":1.7976931348623157e+308},{\"rt\":8,\"r\":1.7976931348623157e+308}]"]
```

File: internal/proxy/multi_rate_limiter.go [hunk 9]
```diff
@@ -49,80 +50,98 @@ func GetQuotaErrorString(errCode commonpb.ErrorCode) string {
...
-func (m *MultiRateLimiter) SetQuotaStates(states []milvuspb.QuotaState, codes []commonpb.ErrorCode) {
+func (m *MultiRateLimiter) SetRates(rates []*proxypb.CollectionRate) error {
 	m.quotaStatesMu.Lock()
 	defer m.quotaStatesMu.Unlock()
-	m.quotaStates = make(map[milvuspb.QuotaState]commonpb.ErrorCode, len(states))
-	for i := 0; i < len(states); i++ {
-		m.quotaStates[states[i]] = codes[i]
+	m.collectionLimiters = make(map[int64]*rateLimiter, 0)
+	for _, collectionRates := range rates {
+		rateLimiter := newRateLimiter()
```
Comment (bigsheeper) [inline internal/proxy/multi_rate_limiter.go:124]: Creating a new limiter for every setting rate may have too much overhead. Maybe we can create a limiter only if it does not already exist, and remove it if it is not in the current collection of rates.

File: internal/proxy/multi_rate_limiter.go (hunk 6 above)
Comment (weiliu1031) [reply internal/proxy/multi_rate_limiter.go:67]: for now, the error code here are expose to sdk side by response, it's hard to warp it, may cause side effect.

File: internal/proxy/multi_rate_limiter.go [hunk 10]
```diff
@@ -138,34 +166,65 @@ func (rl *rateLimiter) limit(rt internalpb.RateType, n int) (bool, float64) {
...
 			return fmt.Errorf("unregister rateLimiter for rateType %s", r.GetRt().String())
 		}
+		log.RatedInfo(30, "current collection rates in proxy",
+			zap.String("rateType", r.Rt.String()),
+			zap.Float64("rateLimit", r.GetR()),
+		)
+	}
+
+	for i := 0; i < len(collectionRate.GetStates()); i++ {
+		rl.quotaStates.Insert(collectionRate.States[i], collectionRate.Codes[i])
```
Comment (bigsheeper) [inline internal/proxy/multi_rate_limiter.go:188]: When are states removed from quotaStates? Perhaps we should create a new quotaStates map before inserting.

File: internal/proxy/multi_rate_limiter.go [hunk 11]
```diff
@@ -138,34 +166,65 @@ func (rl *rateLimiter) limit(rt internalpb.RateType, n int) (bool, float64) {
...
 		if limit, ok := rl.limiters.Get(r.GetRt()); ok {
 			limit.SetLimit(ratelimitutil.Limit(r.GetR()))
-			setRateGaugeByRateType(r.GetRt(), paramtable.GetNodeID(), r.GetR())
+			setRateGaugeByRateType(r.GetRt(), paramtable.GetNodeID(), collectionRate.Collection, r.GetR())
 		} else {
 			return fmt.Errorf("unregister rateLimiter for rateType %s", r.GetRt().String())
 		}
+		log.RatedInfo(30, "current collection rates in proxy",
+			zap.String("rateType", r.Rt.String()),
+			zap.Float64("rateLimit", r.GetR()),
```
Comment (bigsheeper) [inline internal/proxy/multi_rate_limiter.go:183]: May be very long,  how about using `ratelimitutil.Limit(r.GetR()).String()`

File: internal/rootcoord/quota_center.go [hunk 12]
```diff
@@ -183,11 +192,19 @@ func (q *QuotaCenter) syncMetrics() error {
 		if err != nil {
 			return err
 		}
+		for k := range q.queryNodeMetrics {
+			delete(q.queryNodeMetrics, k)
```
Comment (bigsheeper) [inline internal/rootcoord/quota_center.go:196]: With q.clearMetrics, no need to delete here?

File: internal/rootcoord/quota_center.go [hunk 13]
```diff
@@ -204,11 +221,20 @@ func (q *QuotaCenter) syncMetrics() error {
 		if err != nil {
 			return err
 		}
+
+		for k := range q.dataNodeMetrics {
+			delete(q.dataNodeMetrics, k)
```
Comment (bigsheeper) [inline internal/rootcoord/quota_center.go:226]: ditto

File: internal/rootcoord/quota_center.go [hunk 14]
```diff
@@ -290,40 +334,84 @@ func (q *QuotaCenter) calculateReadRates() {
...
+	// query latency
+	queueLatencyThreshold := Params.QuotaConfig.QueueLatencyThreshold.GetAsDuration(time.Second)
+	// queueLatencyThreshold >= 0 means enable queue latency protection
+	if queueLatencyThreshold >= 0 {
+		for _, metric := range q.queryNodeMetrics {
+			searchLatency := metric.SearchQueue.AvgQueueDuration
+			queryLatency := metric.QueryQueue.AvgQueueDuration
+			if searchLatency >= queueLatencyThreshold || queryLatency >= queueLatencyThreshold {
+				collections := metric.Effect.CollectionIDs
+				coolOff(realTimeSearchRate, realTimeQueryRate, collections...)
```
Comment (bigsheeper) [inline internal/rootcoord/quota_center.go:377]: seems the collection will cool-off many times if it is loaded in many query nodes?

File: internal/rootcoord/quota_center.go [hunk 15]
```diff
@@ -290,40 +334,84 @@ func (q *QuotaCenter) calculateReadRates() {
...
+		sum := func(ri metricsinfo.ReadInfoInQueue) int64 {
+			return ri.UnsolvedQueue + ri.ReadyQueue + ri.ReceiveChan + ri.ExecuteChan
+		}
+		for _, metric := range q.queryNodeMetrics {
+			searchNQSum := sum(metric.SearchQueue)
+			queryTasksSum := sum(metric.QueryQueue)
+			nqInQueue := searchNQSum + queryTasksSum // We think of the NQ of query request as 1.
+			if nqInQueue >= nqInQueueThreshold {
+				collections := metric.Effect.CollectionIDs
+				coolOff(realTimeSearchRate, realTimeQueryRate, collections...)
```
Comment (bigsheeper) [inline internal/rootcoord/quota_center.go:395]: ditto

File: internal/rootcoord/quota_center.go [hunk 16]
```diff
@@ -340,281 +428,178 @@ func (q *QuotaCenter) calculateWriteRates() error {
...
-			zap.Duration("MaxDelay", maxDelay),
-			zap.Float64("factor", factor))
-	}
-	return factor
-}
+	for nodeID, metric := range q.queryNodeMetrics {
+		if enableTtProtection && maxDelay >= 0 && metric.Fgm.NumFlowGraph > 0 && metric.Fgm.MinFlowGraphChannel != "" {
+			minTs, _ := tsoutil.ParseTS(metric.Fgm.MinFlowGraphTt)
+			delay := curTs.Sub(minTs)
+			processTtDelay(typeutil.QueryNodeRole, nodeID, metric.Fgm.MinFlowGraphChannel, minTs, delay, metric.Effect.CollectionIDs)
```
Comment (bigsheeper) [inline internal/rootcoord/quota_center.go:511]: ditto

File: internal/rootcoord/quota_center.go [hunk 17]
```diff
@@ -340,281 +428,178 @@ func (q *QuotaCenter) calculateWriteRates() error {
...
+			metrics.RootCoordTtDelay.WithLabelValues(typeutil.QueryNodeRole, strconv.FormatInt(nodeID, 10)).Set(float64(delay.Milliseconds()))
+		}
 
-// getNQInQueryFactor checks search&query nq in QueryNode,
-// and return the factor according to NQInQueueThreshold.
-func (q *QuotaCenter) getNQInQueryFactor() float64 {
-	if !Params.QuotaConfig.QueueProtectionEnabled.GetAsBool() {
-		return 1
+		if enableMemProtection {
+			processMemUsage(typeutil.QueryNodeRole, nodeID, metric.Hms, queryNodeMemoryLowWaterLevel, queryNodeMemoryHighWaterLevel, metric.Effect.CollectionIDs)
```
Comment (bigsheeper) [inline internal/rootcoord/quota_center.go:516]: ditto

File: internal/rootcoord/quota_center.go [hunk 18]
```diff
@@ -340,281 +428,178 @@ func (q *QuotaCenter) calculateWriteRates() error {
...
 	}
 
-	sum := func(ri metricsinfo.ReadInfoInQueue) int64 {
-		return ri.UnsolvedQueue + ri.ReadyQueue + ri.ReceiveChan + ri.ExecuteChan
-	}
+	for nodeID, metric := range q.dataNodeMetrics {
+		if enableTtProtection && maxDelay >= 0 && metric.Fgm.NumFlowGraph > 0 && metric.Fgm.MinFlowGraphChannel != "" {
+			minTs, _ := tsoutil.ParseTS(metric.Fgm.MinFlowGraphTt)
+			delay := curTs.Sub(minTs)
+			processTtDelay(typeutil.DataNodeRole, nodeID, metric.Fgm.MinFlowGraphChannel, minTs, delay, metric.Effect.CollectionIDs)
```
Comment (bigsheeper) [inline internal/rootcoord/quota_center.go:524]: ditto

File: internal/rootcoord/quota_center.go [hunk 19]
```diff
@@ -340,281 +428,178 @@ func (q *QuotaCenter) calculateWriteRates() error {
...
+			metrics.RootCoordTtDelay.WithLabelValues(typeutil.DataNodeRole, strconv.FormatInt(nodeID, 10)).Set(float64(delay.Milliseconds()))
+		}
 
-	nqInQueueThreshold := Params.QuotaConfig.NQInQueueThreshold.GetAsInt64()
-	if nqInQueueThreshold < 0 {
-		// < 0 means disable queue length protection
-		return 1
+		if enableMemProtection {
+			processMemUsage(typeutil.QueryNodeRole, nodeID, metric.Hms, dataNodeMemoryLowWaterLevel, dataNodeMemoryHighWaterLevel, metric.Effect.CollectionIDs)
```
Comment (bigsheeper) [inline internal/rootcoord/quota_center.go:529]: ditto

File: internal/rootcoord/quota_center.go [hunk 20]
```diff
@@ -246,20 +272,36 @@ func (q *QuotaCenter) syncMetrics() error {
...
 
 // forceDenyWriting sets dml rates to 0 to reject all dml requests.
-func (q *QuotaCenter) forceDenyWriting(errorCode commonpb.ErrorCode) {
-	q.currentRates[internalpb.RateType_DMLInsert] = 0
-	q.currentRates[internalpb.RateType_DMLDelete] = 0
-	q.currentRates[internalpb.RateType_DMLBulkLoad] = 0
-	log.Warn("QuotaCenter force to deny writing", zap.String("reason", errorCode.String()))
-	q.quotaStates[milvuspb.QuotaState_DenyToWrite] = errorCode
+func (q *QuotaCenter) forceDenyWriting(errorCode commonpb.ErrorCode, collections ...int64) {
+	if len(collections) == 0 {
```
Comment (bigsheeper) [inline internal/rootcoord/quota_center.go:276]: when `len(collections) == 0` and len(q.writableCollections) != 0 ?

File: internal/rootcoord/quota_center.go (hunk 12 above)
Comment (weiliu1031) [reply internal/rootcoord/quota_center.go:196]: for nodedown case, the nodes list decrease, so we should clear old metrics, not just insert new one

**Pull request merged**